### ISSUE TITLE
Freehand doesn't work with tablets that (brokenly) don't send pressure (Huion P608N, Genius EasyPen i405X, others...?)

### ISSUE THREAD

**Issue by keantoken** [mypaint/mypaint#732]
Freehand won't work but the paint bucket works. It is detecting the tip but for some reason not drawing anything. Everywhere else it works normally.
#### Basic system details

MyPaint version: 1.2.0+gitexport.f62444e
Operating system: Debian Sid
Desktop environment: Xfce
# 

Terminal output:

```
~$ mypaint
INFO: mypaint: Installation layout: conventional POSIX-like structure with prefix u'/usr'
INFO: mypaint: There is an old-style configuration area in u'/home/keantoken/.mypaint/'
INFO: mypaint: Its contents can be migrated to $XDG_CONFIG_HOME and $XDG_DATA_HOME if you wish.
INFO: mypaint: See the XDG Base Directory Specification for info.
INFO: lib.i18n: POSIX: LANG='en_US.utf8'
INFO: lib.i18n: POSIX: LANGUAGE=None
INFO: gui.device: New device 'HUION P608N Mouse' (GDK_SOURCE_MOUSE, axes:4, class=X11DeviceXI2, vendor='256c', product='006e')
INFO: gui.device: New device 'HUION P608N Pen' (GDK_SOURCE_PEN, axes:3, class=X11DeviceXI2, vendor='256c', product='006e')
INFO: gui.device: New device 'Logitech USB-PS/2 Optical Mouse' (GDK_SOURCE_MOUSE, axes:4, class=X11DeviceXI2, vendor='046d', product='c044')
INFO: gui.device: New device 'Virtual core XTEST pointer' (GDK_SOURCE_MOUSE, axes:2, class=X11DeviceXI2, vendor=None, product=None)
INFO: gui.document: Initialized background from u'/usr/share/mypaint/backgrounds/10_soft_yellow.png'
WARNING: gui.keyboard: Ignoring keybinding for '<Actions>/BrushModifierActions/BlendModeMenu'

(mypaint:29741): Gtk-WARNING **: Theme parsing error: <data>:5:50: The style property GtkMenuBar:internal-padding is deprecated and shouldn't be used anymore. It will be removed in a future version
INFO: lib.document: Using u'load_ora' to load u'/home/keantoken/.mypaint/scratchpads/autosave.ora' (kwargs={'feedback_cb': <function gtk_main_tick at 0x7f79de1c4140>})
INFO: lib.document: load_ora: u'/home/keantoken/.mypaint/scratchpads/autosave.ora'

(mypaint:29741): Gtk-WARNING **: Allocating size to GtkSeparatorToolItem 0x37c3560 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(mypaint:29741): Gtk-WARNING **: Allocating size to GtkSeparatorToolItem 0x37c3f80 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?

(mypaint:29741): Gtk-WARNING **: Allocating size to GtkSeparatorToolItem 0x37c59f0 without calling gtk_widget_get_preferred_width/height(). How does the code know the size to allocate?
INFO: lib.document: 0.113s load_ora total
INFO: gui.filehandling: Loaded scratchpad from u'/home/keantoken/.mypaint/scratchpads/autosave.ora'
WARNING: gui.accelmap: Not listing '<Actions>/DocumentActions/RotateLeft' (no tooltip, fix before release!)
WARNING: gui.accelmap: Not listing '<Actions>/DocumentActions/RotateRight' (no tooltip, fix before release!)
```

**Comment by achadwick** [comment]
I assume you mean "everywhere else in MyPaint".

Have you proven that pressure is working at all? This behaviour sounds like the brokenness you get with the Genius EasyPen i405X. Xorg doesn't have decent native support for that, as of unpatched xserver-xorg 7.7+16.

1. Please install `gtk-3-examples`, run *gtk3-demo*, and run the _Event Axes_ demo.

2. Output of `xinput list` please, then `xinput test ID_FROM_LIST_GOES_HERE`, e.g. `xinput test 15`. You should see input from at least one axis >= 2, like `a[2]`, along with the absolute X and Y coord axes.

Annoyingly I've lost the charging thingy for my test Huion H610PRO's pen, so I can't tell you if this is a general Huion regression in Sid. Please test more at your end and tell me what you see.

**Comment by keantoken** [comment]
I have gotten pressure working before in the past, but not with Mypaint. It has worked fine without it.

```
$ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Logitech USB-PS/2 Optical Mouse         	id=10	[slave  pointer  (2)]
⎜   ↳ HUION P608N Pen                         	id=11	[slave  pointer  (2)]
⎜   ↳ HUION P608N Mouse                       	id=12	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Power Button                            	id=7	[slave  keyboard (3)]
    ↳ CHICONY USB Keyboard                    	id=8	[slave  keyboard (3)]
    ↳ CHICONY USB Keyboard                    	id=9	[slave  keyboard (3)]
    ↳ Eee PC WMI hotkeys       
```

From xinput, this is the most I get:

```
motion a[0]=17938 a[1]=22289 
motion a[0]=17917 a[1]=22470 
motion a[0]=17894 a[1]=22766 
```

Maybe something had changed in the Kernel drivers?

**Comment by achadwick** [comment]
I suspect we might have a regression for tablets which don't send a pressure axis. It might be related to the new code that manages events that arrive with spurious zero-pressure axis values when stuff *doesn't* change. As noted in #29 - we often see this sort of nonsense from evdev kit:

![axischart](https://cloud.githubusercontent.com/assets/61299/3563308/f5ab7e36-0a2d-11e4-919f-2399352b6d6b.png)

The horror is real. We deal with this edge case by remembering the last-known-good value and replaying that during motion event handling. There's special handling around button presses and releases to clear out the stored state, and it is _sinfully_ ugly code really -- that's why it might be a source of this glitch.

What you're seeing is presumably a situation where there never was a defined pressure axis value to start with, so this kind of gap-filling won't work.

@keantoken, do you see button press and release events in your `xinput test` stream? If you do, I may be able to use my Genius i405X to emulate some of what you're seeing enough to work out a fix for this.

**Comment by achadwick** [comment]
Better handling would be to assume a constant pressure of 0.5 during motions just like our current mouse event processing.

**Comment by keantoken** [comment]
I see stuff like this when pressing buttons:

```
motion a[0]=11066 a[1]=8599 
motion a[0]=11052 a[1]=8551 
motion a[0]=10986 a[1]=8460 
motion a[0]=10985 a[1]=8457 
motion a[0]=10983 a[1]=8450 
button press   2 
motion a[1]=10620 
motion a[0]=12330 a[1]=10566 
motion a[0]=12336 a[1]=10515 
button press   1 
motion a[0]=12343 a[1]=10468 
motion a[0]=12356 a[1]=10427 
motion a[0]=12375 a[1]=10388 
motion a[0]=12401 a[1]=10350 
motion a[0]=10978 a[1]=8439 
motion a[0]=10969 a[1]=8428 

motion a[1]=6030 
button press   3 
motion a[1]=5977 
motion a[1]=5930 
motion a[1]=5884 
motion a[1]=5842 
motion a[1]=5798 
motion a[1]=5754 
motion a[1]=5714 
motion a[0]=7853 a[1]=5676 
motion a[1]=5637 
motion a[1]=5604 
motion a[1]=5571 
motion a[1]=5541 
motion a[1]=5515 
motion a[0]=7850 a[1]=5491 
motion a[0]=7846 a[1]=5468 
motion a[0]=7840 a[1]=5450 
motion a[0]=7832 a[1]=5434 
motion a[0]=7823 a[1]=5420 
motion a[0]=7813 a[1]=5408 
motion a[0]=7801 a[1]=5395 
motion a[0]=7787 a[1]=5383 
motion a[0]=7771 a[1]=5373 
motion a[0]=7754 a[1]=5368 
motion a[0]=7736 
motion a[0]=7716 a[1]=5340 
button release 3 
motion a[0]=7561 a[1]=5369 

```

Button 1 is the pen tip:

```
motion a[0]=13488 a[1]=9761 
button press   1 
motion a[0]=13554 a[1]=9748 
motion a[0]=13621 a[1]=9737 
motion a[0]=13685 a[1]=9727 
motion a[0]=13748 a[1]=9719 
motion a[0]=13808 a[1]=9713 
motion a[0]=13865 a[1]=9708 
motion a[0]=13921 a[1]=9705 
motion a[0]=13975 
motion a[0]=14031 a[1]=9706 
motion a[0]=14089 
button release 1 
motion a[0]=14151 a[1]=9704 
```

It's a pity, these, tablets were sold at a discount and have reasonably good specs, I thought I was getting a deal. :(

**Comment by achadwick** [comment]
Found the thingy for charging my Huion pen. The H610PRO is still happy, so at least it's not a general driver regression for all Huions.

Looks like it works in a similar way to the Genius tablet I have on the shelf: 

* no pressure axis :cry: 
* absolute (X,Y) axes in motion events
* missing axis data in some motion events, probably when the valuator hasn't changed :fearful: 
* presence of button-1 press and release events for the pen nib :relieved: 

It *should* work in the same way as a mouse or a touchpad, both of which produce identical behaviour in xinput. So why can't these tablets be drawn with? Hmm.

**Comment by achadwick** [comment]
IMO, we should fix up the behaviour for these broken tablets if they can be coaxed into working more like a mouse.

@keantoken I'll look into it - having hardware to test means I actually can, for a change! Would you be up for testing a couple of things?

* Please try the DIGImend drivers https://github.com/DIGImend/digimend-kernel-drivers please, to see if they restore sensible behaviour. Perhaps you had them installed before, had them removed or made useless in a kernel upgrade, and hence saw a regression?
* Please stay subscribed to this bug so I can get you to test any fixes I put together for the mainstream kernel.

**Comment by achadwick** [comment]
**TRIAGE NOTE**: this is an upstream bug in the sense that the tablets do not have pressure available. However MyPaint has bug of its own too relating to the handling of the grotty input. There's clearly something about it that differs from a mouse or a touchpad.

**Comment by achadwick** [comment]
Okay, there's actually not a lot we can do with this in the Freehand code. You remember when I said these devices don't report a pressure axis? Well, as far as xi2 is concerned these tablets _do_ have a pressure axis. That means GTK dutifully reports each event as having a pressure axis, with a value of zero. They also present as tablet devices, just like ones which work. For MyPaint this means that we probably cannot distinguish them from devices which *do* report variations in pressure!

### How to check whether you have of these buggy devices

You can see for yourself with `xinput test-xi2`, which reports a block of information about the device before starting its event reporting. Known valuators are reported there, and pressure is reported as "something the tablet can do" for my i405X:

```
$ xinput list
[...]
⎜   ↳ Genius EasyPen i405X                    	id=14	[slave  pointer  (2)]
[...]
$ xinput test-xi2 14
Genius EasyPen i405X                    	id=14	[slave  pointer  (2)]
	Reporting 4 classes:
[... XIButtonClass omitted ...]
[... "Abs X" (Valuator 0) and "Abs Y" (Valuator 1) XIValuatorClasses omitted ...]
		Class originated from: 14. Type: XIValuatorClass
		Detail for Valuator 2:
		  Label: Abs Pressure
		  Range: 0.000000 - 1023.000000
		  Resolution: 0 units/m
		  Mode: absolute
		  Current value: 0.000000
```

But you don't see any trace of Valuator 2 in the event stream, ever: :sob: 

```
[...]
EVENT type 17 (RawMotion)
    device: 14 (14)
    detail: 0
    flags: 
    valuators:
          0: 9629.00 (9629.00)
          1: 3159.00 (3159.00)
[...]
```

tl;dr - that's how you spot a device which isn't going to cooperate sensibly with MyPaint. Or with *anything* that tries to handle pressure, unless it has a workaround.

**Comment by achadwick** [comment]
One workaround would be for the user to manually tell MyPaint to ignore pressure on the device. Another case for the prefs dialogs "Use for..." menu? We use that for working around certain devices breaking stuff already.

![prefs_50pc](https://cloud.githubusercontent.com/assets/61299/18189192/7e67a96c-70b0-11e6-8c77-c2d1b7d3299f.png)

“Use for: All (constant pressure)?”, perhaps.

Detection of device types is pleasingly good in X11 GTK3, much more so than X11 GTK2. So perhaps the default device-usage you see in that dialog could depend on name and type based heuristics. We could even replace our existing `last_event_had_pressure` hack for detecting mice if we get really lucky.

(It's much clunkier in Windows: everything detects as a mouse over there. So heuristics like these should only apply on `os.name=="posix"` systems)

**Comment by keantoken** [comment]
I did try installing the Digimend and Wizardpen drivers at some point. My tablet did not work right and I carefully constructed a zip file containing data coming directly from the USB port, which did include the pressure data. Should I find the zip file and upload it? I gave the zip file to the driver maintainer but he seemed to be inactive and nothing ever came of it.

Huion released their own version of Linux drivers, surprisingly, and I got the Liquorix maintainer to add them to the kernel, but they are badly written and the cursor was confined to the upper left of the screen.

I just installed the Digimend drivers and nothing has changed.

**Comment by achadwick** [comment]
@keantoken
Not here please - it's irrelevant to MyPaint. That data needs to be sent to either the kernel maintainers, or (probably more productively) the DIGImend crew. Please raise an issue over at https://github.com/DIGImend/digimend-kernel-drivers/issues if it hasn't already been addressed and you want somebody to work on it with you.

By design, MyPaint does not and will not include its own hardware drivers.

**Comment by achadwick** [comment]
Shower thought: is it possible to automatically detect a tablet with broken drivers (always presents pressure=0), and apply a workaround automatically? Presumably normal tablets' drivers will send a defined, positive, non-zero pressure between a pair of button-1-press and button-1-release events...?

(This might be a heuristic too far, and it would lead to oddities like tablets switching behaviour on the fly. Let's try the prefs-based workaround first.)

**Comment by keantoken** [comment]
Well, I tried:

https://github.com/DIGImend/digimend-kernel-drivers/issues/45#issue-174722799

I was not able to get freehand back by changing the device usage page (though I suspect you didn't mean for me to try). I will see if different kernels give different results.

**Comment by achadwick** [comment]
@keantoken Please can you find out:

* Whether your tablet uses the evdev driver. Please look at your system or Xorg logs when you plug in your device, and see if `evdev` is mentioned.
* What version of `xserver-xorg-input-evdev` you have installed. I suspect it's 2.10.2-1 like mine...

From #734, this may be fixed upstream as of 2.10.3:

> Upstream: https://bugs.freedesktop.org/show_bug.cgi?id=95315 (maybe)
> Upstream patch: https://patchwork.freedesktop.org/patch/86593/
> 
> Possibly fixed upstream in: xserver-xorg-input-evdev (Debian) xf86-input-evdev (Arch) version 2.10.3.
> Track changes here: <https://cgit.freedesktop.org/xorg/driver/xf86-input-evdev/log/>.

**Comment by achadwick** [comment]
Since this is looking more and more like an upstream bug (hi @tihoulas, and thanks for the good feedback in #734!), I'm less happy about introducing more hacks to deal with cases that Should Not Happen. Let's wait and see whether updated xf86-input-evdev drivers at version 2.10.3 fix this.

**Comment by achadwick** [comment]
(Unless anyone wants to argue the case for "use for painting, but ignore the pressure" being a useful setting for a tablet...)

**Comment by keantoken** [comment]
Yes, it is the evdev driver and the same version.

**Comment by tihoulas** [comment]
Ok, I built and installed a patched version of xserver-xorg-input-evdev with the upstream patch and it is working now.

[xserver-xorg-input-evdev_2.10.2-1ubuntu1_amd64.deb.zip](https://github.com/mypaint/mypaint/files/467882/xserver-xorg-input-evdev_2.10.2-1ubuntu1_amd64.deb.zip)

**Comment by keantoken** [comment]
Thanks a lot, it works.

**Comment by achadwick** [comment]
Related bug report in Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=825560

**Comment by achadwick** [comment]
New evdev drivers appear to fix this issue. Closing.

**Comment by xiccarph** [comment]
This thread (and fix) worked for me.

I was about to really regret upgrading to Ubuntu 16.10, because it screwed up my ability to use mypaint with my Ugee/UC-Logic 21.5" Tablet Monitor (worked fine on 16.04).  

I just wanted to say thanks to achadwick, whose patient, consistent and superior involvement with the users of my paint makes it a piece of software I enjoy using (and getting to work).  I've spent quite a bit of time looking over various issues here...and I am surprised by how much you engage with everyone.  Thank you.  Thanks also to tihoulas for taking the time to post a patched version.
